src/utils/normalize.ts

```diff
@@ -0,0 +1,94 @@
+import { add0x, isValidHexAddress } from '@metamask/utils';
+import type { Hex } from '@metamask/utils';
+import BN from 'bn.js';
+
+type EIP712Domain = {
+  verifyingContract: string;
+};
+
+type SignTypedMessageDataV3V4 = {
+  types: Record<string, unknown>;
+  domain: EIP712Domain;
+  primaryType: string;
+  message: unknown;
+};
+
+/**
+ * Normalizes the messageData for the eth_signTypedData
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeTypedMessage(messageData: string) {
+  let data;
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
+  } catch (e) {
+    // Ignore normalization errors and pass the message as is
+    return messageData;
+  }
+
+  const { verifyingContract } = data.domain ?? {};
+
+  if (!verifyingContract) {
+    return messageData;
+  }
+
+  data.domain.verifyingContract = normalizeContractAddress(verifyingContract);
+
+  return JSON.stringify(data);
+}
+
+/**
+ * Parses the messageData to obtain the data object for EIP712 normalization
+ *
+ * @param data - The messageData to parse.
+ * @returns The data object for EIP712 normalization.
+ */
+function parseTypedMessage(data: string) {
+  if (typeof data !== 'string') {
+    return data;
+  }
+
+  try {
+    return JSON.parse(data);
+  } catch (e) {
+    throw new Error(`Invalid message data for normalization. data: ${data}`);
+  }
+}
+
+/**
+ * Normalizes the address to a hexadecimal format
+ *
+ * @param address - The address to normalize.
+ * @returns The normalized address.
+ */
+function normalizeContractAddress(address: string): Hex {
+  const addressHex = address as Hex;
+  if (isValidHexAddress(addressHex)) {
+    return addressHex;
+  }
+
+  // Check if the address is in octal format, convert to hexadecimal
+  if (addressHex.startsWith('0o')) {
+    // If octal, convert to hexadecimal
+    return octalToHex(addressHex as string);
+  }
+
+  // Check if the address is in decimal format, convert to hexadecimal
+  const parsedAddress = parseInt(addressHex, 10);
```

Comment:
Do we need to be worried that this will be a large number? I see that you use `BN` below, but you seem to only use it for conversion (and then, only for decimal -> hex and not octal -> hex).

Reply:
Done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/ba7ec73b344b427085f6eabe09045eb824284cc6